File: packages/sdk-snippets/src/targets/python/aws/webhooks/fixtures/security-types.py
import json

import boto3
from readme_metrics.VerifyWebhook import VerifyWebhook

# Your ReadMe secret as a bytes object; you may want to store this in AWS Secrets Manager
README_SECRET = b"my-readme-secret"


def handler(event, lambda_context):
    status_code = None
    email = None
    api_key = None
    error = None

    try:
        signature = event.get("headers", {}).get("ReadMe-Signature")
        body = json.loads(event.get("body", "{}"))
        VerifyWebhook(body, signature, README_SECRET)

        email = body.get("email")
        client = boto3.client("apigateway")
        keys = client.get_api_keys(nameQuery=email, includeValues=True)
        if len(keys.get("items", [])) > 0:
            # if multiple API keys are returned for the given email, use the first one
            api_key = keys["items"][0]["value"]
            status_code = 200
        else:
            error = "Email not found"
            status_code = 404
    except Exception as e:
        error = str(e)
        if error.find("Signature") > -1:
            status_code = 401
        else:
            status_code = 500

    if status_code == 200:
        body = {
            # OAS Security variables
            "api_key": api_key,
            "http_basic": { "user": email, "pass": api_key },
            "http_bearer": api_key,
            "oauth2": api_key,
        }
    else:
        body = {"message": error}

    return {
        "statusCode": status_code,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps(body),
    }

File: packages/sdk-snippets/src/targets/node/aws/webhooks/fixtures/security-types.js
const { APIGatewayClient, GetApiKeysCommand } = require('@aws-sdk/client-api-gateway');
const readme = require('readmeio');

// Your ReadMe secret; you may want to store this in AWS Secrets Manager
const README_SECRET = 'my-readme-secret';

exports.handler = async event => {
  let statusCode, email, apiKey, error;

  try {
    // Verify the request is legitimate and came from ReadMe.
    const signature = event.headers['ReadMe-Signature'];
    const body = JSON.parse(event.body);
    readme.verifyWebhook(body, signature, README_SECRET);

    // Look up the API key associated with the user's email address.
    const email = body.email;
    const client = new APIGatewayClient();
    const command = new GetApiKeysCommand({ nameQuery: email, includeValues: true });
    const keys = await client.send(command);
    if (keys.items.length > 0) {
      // If multiple API keys are returned for the given email, use the first one.
      apiKey = keys.items[0].value;
      statusCode = 200;
    } else {
      error = 'Email not found';
      statusCode = 404;
    }
  } catch (e) {
    error = e.message;
    statusCode = error.match(/Signature/) ? 401 : 500;
  }

  return {
    statusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      // OAS Security variables
      api_key: apiKey,
      http_basic: { user: email, pass: apiKey },
      http_bearer: apiKey,
      oauth2: apiKey,

      // Error message, if any
      message: error,
    }),
  };
};